pkg/common/workbuilder/workbuilder.go

```diff
@@ -0,0 +1,318 @@
+const DefaultManifestThreshold = 0.8
+
+type manifestKey struct {
+	gvk             schema.GroupVersionKind
+	namespace, name string
+}
+
+type requiredManifestMapper map[manifestKey]workapiv1.Manifest
+
+// GetExistingManifestWorks is to get existing manifestWorks which need to be updated.
+type GetExistingManifestWorks func() ([]workapiv1.ManifestWork, error)
+
+// GenerateManifestWork is to generate a new manifestWork with empty manifest, need to make sure the name is unique.
+type GenerateManifestWork func(index int) workapiv1.ManifestWork
+
+type WorkBuilder struct {
+	manifestLimit            int
+	generateManifestWork     GenerateManifestWork
+	getExistingManifestWorks GetExistingManifestWorks
+	workApplier              *workapplier.WorkApplier
+}
+
+func NewWorkBuilder() *WorkBuilder {
+	return &WorkBuilder{
+		manifestLimit: int(float64(DefaultManifestLimit) * DefaultManifestThreshold),
+	}
+}
+
+// WithManifestsLimit is to set the total limit size of manifests in manifestWork. the unit of manifestsLimit is byte.
+// the actual size of manifests will be the 80% of the limit.
+// This is an optional setting, and the default limit is 50k.
+func (w *WorkBuilder) WithManifestsLimit(manifestsLimit int) *WorkBuilder {
+	w.manifestLimit = int(float64(manifestsLimit) * DefaultManifestThreshold)
+	return w
+}
+
+// WithGenerateWorkFunc is to set a Function to generate an empty manifestWork with meta info.
+// This is a required setting.
+func (w *WorkBuilder) WithGenerateWorkFunc(generateManifestWork GenerateManifestWork) *WorkBuilder {
+	w.generateManifestWork = generateManifestWork
+	return w
+}
+
+// WithGetExistingWorksFunc is to set a Function to get the existing manifestWorks.
+// This is a required setting for BuildAndApply.
+func (w *WorkBuilder) WithGetExistingWorksFunc(getExistingWorks GetExistingManifestWorks) *WorkBuilder {
+	w.getExistingManifestWorks = getExistingWorks
+	return w
+}
+
+// WithWorkApplier is to set the workApplier used to apply the manifestWorks.
+// This is a required setting for BuildAndApply.
+func (w *WorkBuilder) WithWorkApplier(workApplier *workapplier.WorkApplier) *WorkBuilder {
+	w.workApplier = workApplier
+	return w
+}
+
+// NewBuild is to build a set of manifestWorks using the objects.
+func (w *WorkBuilder) NewBuild(objects []runtime.Object) ([]workapiv1.ManifestWork, error) {
+	var manifests []workapiv1.Manifest
+
+	for _, object := range objects {
+		manifest, err := buildManifest(object)
+		if err != nil {
+			return nil, err
+		}
+		manifests = append(manifests, manifest)
+	}
+	if len(manifests) == 0 {
+		return nil, fmt.Errorf("there is no manifest")
+	}
+
+	return w.newManifestWorks(manifests, 0)
+}
+
+// BuildAndApply is to build a set manifestWorks using the objects and update the existing manifestWorks.
+func (w *WorkBuilder) BuildAndApply(ctx context.Context, objects []runtime.Object) ([]workapiv1.ManifestWork, error) {
```

Comment:
func (w *WorkBuilder) BuildAndApply(ctx context.Context, objects []runtime.Object, opts....) {
   builder := internalBuilder()
   for _, opts := range opts {
      builder = opts(builder)
  }

  appliedWorks, deletedWorks, err := builder.buildManifestWorks(objects)
}

ExistingWorkOptions(existingWork ) *BuilderOption {
   return func(builder InternalBuilder) InternalBuilder {
     builder.existingWork = existingWork
     return builder
} 

DeleteOption()

FeedBackOption()

Reply:
done